quinn-proto/src/connection.rs

```diff
@@ -301,10 +319,38 @@ impl Connection {
         {
             streams.insert(
                 StreamId::new(!side, Directionality::Bi, i as u64),
-                Stream::new_bi(config.stream_receive_window as u64),
+                Stream::new_bi(ctx.config.stream_receive_window as u64),
             );
         }
-        Self {
+        let (tls, tls_client_config) = match conn_config {
```

Comment:
Hmm, it feels like this chunk of code is more tightly coupled to `Endpoint` than `Connection`, perhaps?

Reply:
How do you mean? The TLS session is certainly part of the connection.

Reply:
I mean, you have a 40-line chunk of code here that transforms a `ConnectionConfig` into a `TlsSession` + `Option<ClientConfig>`. Shouldn't we do that transformation in `Endpoint::add_connection()` instead of inside `Connection`? In that case I think `Connection::new()` wouldn't need access to `ctx` instead of `config`.

And would it make more sense to save the `server_name` in the `ClientConfig`, instead of wrapping another type around it?

Reply:
> Shouldn't we do that transformation in Endpoint::add_connection() instead of inside Connection? In that case I think Connection::new() wouldn't need access to ctx instead of config.

My thinking here was that future work could get rid of `connect` and `handle_initial` entirely, though it's not certain whether that would work out well. I'll try moving things out into `Endpoint` tomorrow and see how that looks.

> And would it make more sense to save the server_name in the ClientConfig, instead of wrapping another type around it?

The `ClientConfig` here is `rustls`'s; `-proto` doesn't have its own `ClientConfig` at the moment, though I've been thinking it could make sense for one to exist eventually; various parameters taken from the endpoint configuration don't actually need to be global across outgoing connections.

Reply:
Oh, yes, I forgot that this was rustls's `ClientConfig`. I'll be curious to see your `Endpoint` variant; I do agree with working towards the elimination of `connect()` and `handle_initial()`, but I'm not sure if that invalidates the point that the `Endpoint` should take care of this bit. Anyway, let's see how it turns out.